lib/ExpensiMark.js

```diff
@@ -137,4 +149,46 @@ export default class ExpensiMark {
 
         return replacedText;
     }
+
+    /**
+     * Checks matched URLs for validity and replace valid links with html elements
+     *
+     * @param {RegExp} regex
+     * @param {String} textToCheck
+     * @param {Function} replacement
+     *
+     * @returns {String}
+     */
+    modifyTextForUrlLinks(regex, textToCheck, replacement) {
+        let match = regex.exec(textToCheck);
+        let replacedText = '';
+        let startIndex = 0;
+        let abort = false;
+
+        while (match !== null) {
+            // we want to avoid matching email address domains
+            if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
+                abort = true;
+            }
+
+            // we want to avoid matching ending ) unless it is a closing parenthesis for the URL
+            if (textToCheck[(match.index + match[2].length) - 1] === ')' && !match[2].includes('(')) {
+                match[0] = match[0].substr(0, match[0].length - 1);
+                match[2] = match[2].substr(0, match[2].length - 1);
+            }
+
+            replacedText = replacedText.concat(textToCheck.substr(startIndex, (match.index - startIndex)));
+
+            if (abort) {
+                replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
+            } else {
+                replacedText = replacedText.concat(replacement(match[0], match[1], match[2], match[3]));
+            }
+            startIndex = match.index + (match[0].length);
+            match = regex.exec(textToCheck);
```

Comment:
NAB: a small comment about how this moves to the next match would be good. Its not obvious if you don't understand how JS's regex engine runs.